android/modules/geolocation/src/java/ti/modules/titanium/geolocation/android/FusedLocationProvider.java

```diff
@@ -0,0 +1,281 @@
+/**
+ * Appcelerator Titanium Mobile
+ * Copyright (c) 2017 by Axway, Inc. All Rights Reserved.
+ * Licensed under the terms of the Apache Public License
+ * Please see the LICENSE included with this distribution for details.
+ */
+package ti.modules.titanium.geolocation.android;
+
+import ti.modules.titanium.geolocation.GeolocationModule;
+import ti.modules.titanium.geolocation.android.LocationProviderProxy.LocationProviderListener;
+
+import android.annotation.SuppressLint;
+import android.content.Context;
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.google.android.gms.common.ConnectionResult;
+import com.google.android.gms.common.GoogleApiAvailability;
+import com.google.android.gms.common.api.ApiException;
+import com.google.android.gms.common.api.GoogleApiClient;
+import com.google.android.gms.location.FusedLocationProviderClient;
+import com.google.android.gms.location.LocationAvailability;
+import com.google.android.gms.location.LocationCallback;
+import com.google.android.gms.location.LocationRequest;
+import com.google.android.gms.location.LocationResult;
+import com.google.android.gms.location.LocationServices;
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.Task;
+
+import org.appcelerator.kroll.common.Log;
+import org.appcelerator.titanium.TiC;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+
+/**
+ * FusedLoactionProvider is used to expose the Google Play Services location API.
+ * However, this functionality is only exposed if the Google Play Services library is detected.
+ */
+public class FusedLocationProvider
+{
+
+	private static final String TAG = "FusedLocationProvider";
+
+	public static final String PROVIDER = "fused";
+
+	private final GeolocationModule geolocationModule;
+
+	/**
+	 * Constructor
+	 * @param context: context to be used when accessing Google APIs
+	 * @param geolocationModule: reference of the geolocation module to obtain accuracy when
+	 *					         registering the location provider.
+	 */
+	public FusedLocationProvider(Context context, GeolocationModule geolocationModule)
+	{
+		this.geolocationModule = geolocationModule;
+
+		if (hasPlayServices()) {
+			PlayServices.init(context, geolocationModule);
+		}
+	}
+
+	/**
+	 * hasPlayServices
+	 * @return do we have access to Google Play Services APIs?
+	 */
+	public static boolean hasPlayServices()
+	{
+		try {
+			Class.forName("com.google.android.gms.common.GoogleApiAvailability");
+		} catch (ClassNotFoundException e) {
+			return false;
+		}
+		return PlayServices.validVersion();
```

Comment:
We should also check if `isGooglePlayServicesAvailable()` returns `SUCCESS` in the `hasPlayServices()` method.

This is to handle the case where `ti.playservices` module is included, but we're running on a device/emulator that doesn't have Google Play Services installed, is disabled, or is too old.